dune/grid/common/p2pcommunicator.hh

```diff
@@ -0,0 +1,240 @@
+/*
+  Copyright 2015 IRIS AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifndef DUNE_COMMUNICATOR_HEADER_INCLUDED
+#define DUNE_COMMUNICATOR_HEADER_INCLUDED
+
+#include <vector>
+#include <set>
+#include <map>
+
+#include <dune/common/parallel/mpihelper.hh>
+#include <dune/common/parallel/collectivecommunication.hh>
+
+// the following implementation is only available in case MPI is available
+#if HAVE_MPI
+#include <dune/common/parallel/mpicollectivecommunication.hh>
+#endif
+
+namespace Dune
+{
+  class SimpleMessageBuffer
+  {
+    typedef std::vector< char >  BufferType;
+
+    mutable BufferType buffer_;
+    const double factor_;
+    mutable size_t pos_;
+public:
+    /** \brief constructor taking memory reserve estimation factor (default is 1.1, i.e. 10% over estimation )
+     */
+    SimpleMessageBuffer( const double factor = 1.1 )
+      : buffer_(), factor_( factor )
+    {
+      resetReadPosition();
+    }
+
+    /** \brief clear the buffer */
+    void clear() { BufferType().swap( buffer_ ); resetReadPosition(); }
```

Comment:
Why is the "hard" clear necessary for buffer_? I would think a plain buffer_.clear() would be enough?